logs_sdk/lib/opentelemetry/sdk/logs/logger_provider.rb

```diff
@@ -0,0 +1,129 @@
+  module SDK
+    module Logs
+      # The SDK implementation of OpenTelemetry::Logs::LoggerProvider.
+      class LoggerProvider < OpenTelemetry::Logs::LoggerProvider
+        attr_reader :resource, :log_record_processors
+
+        EMPTY_NAME_ERROR = 'LoggerProvider#logger called without '\
+          'providing a logger name.'
+        FORCE_FLUSH_ERROR = 'unexpected error in ' \
+          'OpenTelemetry::SDK::Logs::LoggerProvider#force_flush'
+
+        # Returns a new LoggerProvider instance.
+        #
+        # @param [optional Resource] resource The resource to associate with
+        #   new LogRecords created by {Logger}s created by this LoggerProvider.
+        # @param [optional Array] log_record_processors The
+        #   {LogRecordProcessor}s to associate with this LoggerProvider.
+        #
+        # @return [OpenTelemetry::SDK::Logs::LoggerProvider]
+        def initialize(
+          resource: OpenTelemetry::SDK::Resources::Resource.create,
+          log_record_processors: []
+        )
+          @log_record_processors = log_record_processors
+          @mutex = Mutex.new
+          @resource = resource
+          @stopped = false
+        end
+
+        # Creates an {OpenTelemetry::SDK::Logs::Logger} instance.
+        #
+        # @param [optional String] name Instrumentation package name
+        # @param [optional String] version Instrumentation package version
+        #
+        # @return [OpenTelemetry::SDK::Logs::Logger]
+        def logger(name = nil, version = nil)
+          name ||= ''
+          version ||= ''
+
+          OpenTelemetry.logger.warn(EMPTY_NAME_ERROR) if name.empty?
+
+          @mutex.synchronize do
+            OpenTelemetry::SDK::Logs::Logger.new(name, version, self)
+          end
+        end
+
+        # Adds a new log record processor to this LoggerProvider's
+        # log_record_processors.
+        #
+        # @param [LogRecordProcessor] log_record_processor The
+        #   {LogRecordProcessor} to add to this LoggerProvider.
+        def add_log_record_processor(log_record_processor)
+          @mutex.synchronize do
+            @log_record_processors = log_record_processors.dup.push(log_record_processor)
+          end
+        end
+
+        # Attempts to stop all the activity for this LoggerProvider. Calls
+        # {LogRecordProcessor#shutdown} for all registered {LogRecordProcessor}s.
+        #
+        # This operation may block until all log records are processed. Must
+        # be called before turning off the main application to ensure all data
+        # are processed and exported.
+        #
+        # After this is called all newly created {LogRecord}s will be no-op.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
```

Comment:
See: https://opentelemetry.io/docs/specs/otel/logs/sdk/#shutdown 

* Only called once by setting `@stopped` to `true` L92
* Provides a way to let the caller know whether it succeeded, failed, or timed out by returning  `OpenTelemetry::SDK::Logs::EXPORT` constants (L81, L87, L93)
* Implements a timeout through calls to `OpenTelemetry::Common::Utilities.timeout_timestamp` and `OpenTelemetry::Common::Utilities.maybe_timeout` (L84, L86)
* Invokes shutdown on all the processors (L85 - 89)